src/server/pfs/server/driver.go

```diff
@@ -1972,76 +1935,122 @@ func (d *driver) deleteAll(ctx context.Context) error {
 	return nil
 }
 
-func (d *driver) applyWrites(resp *etcd.GetResponse, tree hashtree.OpenHashTree) error {
+// Put the tree into the blob store
+// Only write the records to etcd if the commit does exist and is open.
+// To check that a key exists in etcd, we assert that its CreateRevision
+// is greater than zero.
+func (d *driver) upsertPutFileRecords(ctx context.Context, file *pfs.File, newRecords *pfs.PutFileRecords) error {
+	prefix, err := d.scratchFilePrefix(ctx, file)
+	if err != nil {
+		return err
+	}
+
+	_, err = col.NewSTM(ctx, d.etcdClient, func(stm col.STM) error {
+		revision := stm.Rev(d.openCommits.Path(file.Commit.ID))
+		if revision == 0 {
+			return fmt.Errorf("commit %v is not open", file.Commit.ID)
+		}
```

Comment:
Did you check to make sure that the first write to the system doesn't get recorded as revision `0`?

Reply:
I did not. I think this is analogous to the previous incantation of the code, [seen here](https://github.com/pachyderm/pachyderm/pull/2599/files/f1cccb143f94c3b57ab97dd0e3e5bc90a39d5eb4#diff-4448519cd42a9ad4ceeaeb1804180343L1472)

But I can double check this.

Reply:
Capturing our discussion offline ...

Reasoning about this code is tricky. From manual testing w etcdctl its hard to check a revision of a key. So instead we opted to check that the value here is non empty and the commit ID matches the current commit. Really that's the simplest check we can think of to guarantee the value exists (and isn't an empty struct unmarshalled from an empty string).